src/sdk/python/tests/iotea/core/test_protocol_gateway.py

```diff
@@ -0,0 +1,251 @@
+    test_case.assertEqual(2, protocol_gateway.adapters[0].instance.publish.call_count)
+    protocol_gateway.adapters[0].instance.publish.assert_called_with(test_topic, test_message, publish_options)
+
+
+@pytest.mark.asyncio
+async def test_subscribe_options(test_case, mocker, protocol_gateway):
+    test_topic = 'test'
+
+    # pylint: disable=unused-argument
+    async def mock_subscribe(topic, cb, subscribe_options=None):
+        pass
+
+    mocker.patch('src.iotea.core.util.mqtt_client.MqttProtocolAdapter.subscribe', wraps=mock_subscribe)
+    # test with None sub options
+    await protocol_gateway.subscribe(test_topic, lambda: {}, subscribe_options=None)
+    test_case.assertEqual(1, protocol_gateway.adapters[0].instance.subscribe.call_count)
+    protocol_gateway.adapters[0].instance.subscribe.assert_called_with(test_topic, unittest.mock.ANY, unittest.mock.ANY)
+    args, _ = protocol_gateway.adapters[0].instance.subscribe.call_args
+    subscribe_options = args[2]
+    test_case.assertIsNotNone(subscribe_options)
+
+    # test with custom options: subscribe with platform protocol adapter
+    subscribe_options = ProtocolGateway.create_subscribe_options(True, None)
+    await protocol_gateway.subscribe(test_topic, lambda: {}, subscribe_options=subscribe_options)
+    test_case.assertEqual(2, protocol_gateway.adapters[0].instance.subscribe.call_count)
+    protocol_gateway.adapters[0].instance.subscribe.assert_called_with(test_topic, unittest.mock.ANY, subscribe_options)
+
+
+def test_validate_configuration():
+    with pytest.raises(Exception):
+        ProtocolGateway.validate_configuration('bad configuration')
+
+    # adapters expected in configuration
+    with pytest.raises(Exception):
+        ProtocolGateway.validate_configuration({'no-adapters': 'present'})
+
+    # adapters must be a list
+    with pytest.raises(Exception):
+        ProtocolGateway.validate_configuration({'adapters': 'is not a list'})
+
+    # more than one platform adapter not allowed
+    with pytest.raises(Exception):
+        ProtocolGateway.validate_configuration({'adapters': [{'platform': True}, {'platform': True}]})
+
+    # no platform adapter available, while required
+    with pytest.raises(Exception):
+        ProtocolGateway.validate_configuration({'adapters': [{'platform': False}, {'no-platform-tag': True}]},
+                                               use_platform_protocol_only=True)
+
+
+@pytest.mark.asyncio
+async def test_validate_platform_protocol_usage():
+    mqtt_def_config = MqttProtocolAdapter.create_default_configuration(is_platform_protocol=True)
+    protocol_gateway_config = ProtocolGateway.create_default_configuration([mqtt_def_config])
+    protocol_gateway = ProtocolGateway(protocol_gateway_config, 'TestProtocolGateway', use_platform_protocol_only=True)
+    publish_options = ProtocolGateway.create_publish_options(platform_protocol_only=False)
+    with pytest.raises(Exception):
```

Comment:
Same as above - check for the actual exception raised

Reply:
Added message checks.